src/chrdev.rs

```diff
@@ -30,24 +49,136 @@ impl Builder {
             )
         };
         if res != 0 {
-            return Err(error::Error::from_kernel_errno(res));
+            return Err(Error::from_kernel_errno(res));
+        }
+
+        let mut cdevs = vec![unsafe { mem::zeroed() }; self.file_ops.len()].into_boxed_slice();
+        for (i, file_op) in self.file_ops.iter().enumerate() {
+            unsafe {
+                bindings::cdev_init(&mut cdevs[i], &file_op.0);
+                cdevs[i].owner = &mut bindings::__this_module;
+                let rc = bindings::cdev_add(&mut cdevs[i], dev + i as bindings::dev_t, 1);
+                if rc != 0 {
+                    // TODO: need to clean up the region + cdevs here.
+                    return Err(Error::from_kernel_errno(rc));
+                }
+            }
         }
+
         return Ok(Registration {
             dev,
             count: self.minors.len(),
+            _cdevs: cdevs,
         });
     }
 }
 
 pub struct Registration {
     dev: bindings::dev_t,
     count: usize,
+    _cdevs: Box<[bindings::cdev]>,
 }
 
+unsafe impl Sync for Registration {}
+
 impl Drop for Registration {
     fn drop(&mut self) {
         unsafe {
             bindings::unregister_chrdev_region(self.dev, self.count as _);
```

Comment:
Don't you need to call `cdev_del` here?

More generally, this plus the TODO sort of implies to me that we should have a safe wrapper type around cdev that implements `Drop`.